src/preamble/02_what_is_a_module_and_a_package.md

```diff
@@ -1 +1,366 @@
-# What is a module and a package?
+# What is a module, a component, and a package?
+
+## TL;DR
+
+There is an ongoing initiative to standardize the terminology we use to refer to
+the hierarchy of collections of code in the Haskell community. If you are
+interested in reading more about the you can find a more thorough treatment
+[here](https://gitlab.haskell.org/ghc/ghc/-/wikis/commentary/compiler/units#background).
+
+The most significant concepts in this chapter all refer to ways of dividing and
+aggregating code. They form a hierarchy, and are somewhat defined in terms of
+each other. Here is a list of them in order from smallest to largest are:
+
+1. **Module** - The smallest unit of code defined by GHC and the Haskell
+   language specification.
+2. **Component** - A component is a collection of modules. Usually a `library`,
+   `executable`, or a `test-suite`.
+3. **Unit** - Units are compiled components, while not technically _larger_ than
+   a component, they are further along in the packaging process and therefore
+   closer to a distributable artefact.
+4. **Package** - A package is a collection of components. In the context of
+   cabal a package is identified by a `<package-name>.cabal` file. A package is
```

Comment:
comma after 'cabal', but I actually don't understand why we need to qualify this to that context. What other contexts are there?

Reply:
I think within the context of stack a package is identified by a `stack.yaml` file. But perhaps you are right, the verbage is confusing.

Reply:
Oh, stack! I knew I'm missing something. Are there really stack-only packages with no Hackage counterparts? Or with .cabal files generated from stack.yaml?

Reply:
I am not sure, I did change the text, but the intention behind the original wording was to remind the user that we are talking about cabal rather than stack or even ghc (if the user was confused about the fundamental mechanism).

Reply:
Sure, but GHC doesn't have packages at all and, I just don't know, but probably stack has exactly the same packages (with extra meta-data). Correct me if I'm wrong.

Reply:
OK, I am wrong. GHC has package and package environments and ghc-pkg. So it's probably worth mentioning we don't mean those. Or say somewhere something like "package concept, which coincides, but is not identical to the concept of GHC packages".

Reply:
No no, `stack.yaml` is really for a Stack project. In it we have a `packages` entry that list the different packages. These packages are defined using `.cabal` files as usual (that can be generated from a `package.yaml` but here we don't care).

I wouldn't mention "GHC packages" or just in passing. "GHC packages" are in fact "units" and "package databases" are in fact "unit databases" so it's quite confusing...

Reply:
@hsyl20: agreed on both counts. I'd mention GHC packages only just enough to avoid a vague "cabal context" qualification and definitely avoid mentioning units at all.